ci: go-tests sometimes fail with "TempDir RemoveAll cleanup: [...] directory not empty"

for example:
https://github.com/vocdoni/vocdoni-node/actions/runs/3137078090/jobs/5094795954
```
--- FAIL: TestEntityList (12.05s)
    scrutinizer_test.go:68: list is empty
    scrutinizer_test.go:68: list is empty
    scrutinizer_test.go:68: list is empty
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestEntityList3644702761: directory not empty
FAIL
coverage: 57.1% of statements
FAIL	go.vocdoni.io/dvote/vochain/scrutinizer	37.815s
```

this is due to some goroutine writing to a tempfile, after the calling function returned. being a race, it only happens once in a while.
see https://github.com/golang/go/issues/43547
maybe @mvdan has a trivial solution?

--- a/vochain/app.go
+++ b/vochain/app.go
@@ -91,7 +91,11 @@ func TestBaseApplication(tb testing.TB) *BaseApplication {
 	app.SetTestingMethods()
 
 	// TODO: should this be a Close on the entire BaseApplication?
-	tb.Cleanup(func() { app.State.Close() })
+	tb.Cleanup(func() {
+		if err := app.State.Close(); err != nil {
+			tb.Error(err)
+		}
+	})
 	return app
 }
 

--- a/vochain/scrutinizer/db_benchmark_test.go
+++ b/vochain/scrutinizer/db_benchmark_test.go
@@ -39,10 +39,7 @@ func BenchmarkNewProcess(b *testing.B) {
 func benchmarkIndexTx(b *testing.B) {
 	app := vochain.TestBaseApplication(b)
 
-	sc, err := NewScrutinizer(b.TempDir(), app, true)
-	if err != nil {
-		b.Fatal(err)
-	}
+	sc := newTestScrutinizer(b, app, true)
 	pid := util.RandomBytes(32)
 	if err := app.State.AddProcess(&models.Process{
 		ProcessId:    pid,
@@ -55,7 +52,7 @@ func benchmarkIndexTx(b *testing.B) {
 	}); err != nil {
 		b.Fatal(err)
 	}
-	err = sc.newEmptyProcess(pid)
+	err := sc.newEmptyProcess(pid)
 	qt.Assert(b, err, qt.IsNil)
 
 	s := new(ethereum.SignKeys)
@@ -83,17 +80,14 @@ func benchmarkFetchTx(b *testing.B) {
 	numTxs := 1000
 	app := vochain.TestBaseApplication(b)
 
-	sc, err := NewScrutinizer(b.TempDir(), app, true)
-	if err != nil {
-		b.Fatal(err)
-	}
+	sc := newTestScrutinizer(b, app, true)
 
 	for i := 0; i < b.N; i++ {
 		sc.Rollback()
 		for j := 0; j < numTxs; j++ {
 			sc.OnNewTx([]byte(fmt.Sprintf("hash%d%d", i, j)), uint32(i), int32(j))
 		}
-		err = sc.Commit(uint32(i))
+		err := sc.Commit(uint32(i))
 		qt.Assert(b, err, qt.IsNil)
 
 		time.Sleep(time.Second * 2)
@@ -118,10 +112,7 @@ func benchmarkFetchTx(b *testing.B) {
 func benchmarkNewProcess(b *testing.B) {
 	app := vochain.TestBaseApplication(b)
 
-	sc, err := NewScrutinizer(b.TempDir(), app, true)
-	if err != nil {
-		b.Fatal(err)
-	}
+	sc := newTestScrutinizer(b, app, true)
 	startTime := time.Now()
 	numProcesses := b.N
 	entityID := util.RandomBytes(20)
@@ -140,7 +131,7 @@ func benchmarkNewProcess(b *testing.B) {
 		}); err != nil {
 			b.Fatal(err)
 		}
-		err = sc.newEmptyProcess(pid)
+		err := sc.newEmptyProcess(pid)
 		qt.Assert(b, err, qt.IsNil)
 
 	}

--- a/vochain/scrutinizer/process.go
+++ b/vochain/scrutinizer/process.go
@@ -5,6 +5,7 @@ import (
 	"flag"
 	"fmt"
 	"strings"
+	"sync/atomic"
 	"time"
 
 	"github.com/google/go-cmp/cmp"
@@ -363,10 +364,17 @@ func (s *Scrutinizer) isOpenProcess(processID []byte) (bool, error) {
 
 // compute results if the current heigh has scheduled ending processes
 func (s *Scrutinizer) computePendingProcesses(height uint32) {
+	defer atomic.AddInt64(&s.liveGoroutines, -1)
 	// We wait a random number of blocks (between 0 and 5) in order to decrease the collision risk
 	// between several Oracles.
 	targetHeight := s.App.Height() + uint32(util.RandomInt(0, 5))
-	for s.App.Height() < targetHeight {
+	for !s.skipTargetHeightSleeps {
+		if s.cancelCtx.Err() != nil {
+			return // closing
+		}
+		if s.App.Height() >= targetHeight {
+			break
+		}
 		time.Sleep(5 * time.Second)
 	}
 	if enableBadgerhold {

--- a/vochain/scrutinizer/scrutinizer.go
+++ b/vochain/scrutinizer/scrutinizer.go
@@ -10,6 +10,7 @@ import (
 	"math/big"
 	"os"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/dgraph-io/badger/v3"
@@ -103,6 +104,18 @@ type Scrutinizer struct {
 	recoveryBootLock sync.RWMutex
 	// ignoreLiveResults if true, partial/live results won't be calculated (only final results)
 	ignoreLiveResults bool
+
+	liveGoroutines int64 // atomic
+
+	// In the tests, the extra 5s sleeps can make CI really slow at times, to
+	// the point that it times out. Skip that in the tests.
+	skipTargetHeightSleeps bool
+
+	// Note that cancelling currently only stops asynchronous goroutines started
+	// by Commit. In the future we could make it stop all other work as well,
+	// like entire calls to Commit.
+	cancelCtx  context.Context
+	cancelFunc context.CancelFunc
 }
 
 // VoteWithIndex holds a Vote and a txIndex. Model for the VotePool.
@@ -115,7 +128,14 @@ type VoteWithIndex struct {
 // NewScrutinizer returns an instance of the Scrutinizer
 // using the local storage database of dbPath and integrated into the state vochain instance
 func NewScrutinizer(dbPath string, app *vochain.BaseApplication, countLiveResults bool) (*Scrutinizer, error) {
-	s := &Scrutinizer{App: app, ignoreLiveResults: !countLiveResults}
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	s := &Scrutinizer{
+		App:               app,
+		ignoreLiveResults: !countLiveResults,
+
+		cancelCtx:  cancelCtx,
+		cancelFunc: cancelFunc,
+	}
 	var err error
 	s.db, err = InitDB(dbPath)
 	if err != nil {
@@ -158,6 +178,42 @@ func NewScrutinizer(dbPath string, app *vochain.BaseApplication, countLiveResult
 	s.envelopeHeightCache = lru.New(countEnvelopeCacheSize)
 	s.resultsCache = lru.New(resultsCacheSize)
 	return s, nil
+}
+
+func (s *Scrutinizer) Close() error {
+	s.cancelFunc()
+	s.WaitIdle()
+	// Try closing both before reporting errors.
+	err1 := s.db.Close()
+	err2 := s.sqlDB.Close()
+	if err1 != nil {
+		return err1
+	}
+	if err2 != nil {
+		return err2
+	}
+	return nil
+}
+
+// WaitIdle waits until there are no live asynchronous goroutines, such as those
+// started by the Commit method.
+//
+// Note that this method is racy by nature if the scrutinizer keeps committing
+// more blocks, as it simply waits for any goroutines already started to stop.
+func (s *Scrutinizer) WaitIdle() {
+	var slept time.Duration
+	for {
+		const debounce = 100 * time.Millisecond
+		time.Sleep(debounce)
+		slept += debounce
+		if atomic.LoadInt64(&s.liveGoroutines) == 0 {
+			break
+		}
+		if slept > 10*time.Second {
+			log.Warnf("giving up on WaitIdle after 10s")
+			return
+		}
+	}
 }
 
 func (s *Scrutinizer) timeoutQueries() (*scrutinizerdb.Queries, context.Context, context.CancelFunc) {
@@ -368,6 +424,7 @@ func (s *Scrutinizer) Commit(height uint32) error {
 	}
 
 	// Index new transactions
+	atomic.AddInt64(&s.liveGoroutines, 1)
 	go s.indexNewTxs(s.newTxPool)
 
 	// Schedule results computation
@@ -472,6 +529,7 @@ func (s *Scrutinizer) Commit(height uint32) error {
 	// an initial results height of 0, and we don't want to compute results
 	// for such an initial height.
 	if height > 0 {
+		atomic.AddInt64(&s.liveGoroutines, 1)
 		go s.computePendingProcesses(height)
 	}
 	return nil

--- a/vochain/scrutinizer/scrutinizer_test.go
+++ b/vochain/scrutinizer/scrutinizer_test.go
@@ -8,7 +8,6 @@ import (
 	stdlog "log"
 	"math/big"
 	"testing"
-	"time"
 
 	qt "github.com/frankban/quicktest"
 	"github.com/pressly/goose/v3"
@@ -28,18 +27,31 @@ func init() {
 	goose.SetLogger(stdlog.New(io.Discard, "", 0))
 }
 
+func newTestScrutinizer(tb testing.TB, app *vochain.BaseApplication, countLiveResults bool) *Scrutinizer {
+	sc, err := NewScrutinizer(tb.TempDir(), app, true)
+	if err != nil {
+		tb.Fatal(err)
+	}
+	sc.skipTargetHeightSleeps = true
+	tb.Cleanup(func() {
+		if err := sc.Close(); err != nil {
+			tb.Error(err)
+		}
+	})
+	return sc
+}
+
 func TestEntityList(t *testing.T) {
-	testEntityList(t, 2)
-	testEntityList(t, 100)
-	testEntityList(t, 155)
+	for _, count := range []int{2, 100, 155} {
+		t.Run(fmt.Sprintf("count=%03d", count), func(t *testing.T) {
+			testEntityList(t, count)
+		})
+	}
 }
 
 func testEntityList(t *testing.T, entityCount int) {
 	app := vochain.TestBaseApplication(t)
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 	for i := 0; i < entityCount; i++ {
 		pid := util.RandomBytes(32)
 		if err := app.State.AddProcess(&models.Process{
@@ -76,16 +88,13 @@ func testEntityList(t *testing.T, entityCount int) {
 		last += 10
 	}
 	if len(entities) < entityCount {
-		t.Fatalf("expected %d entityes, got %d", entityCount, len(entities))
+		t.Fatalf("expected %d entities, got %d", entityCount, len(entities))
 	}
 }
 
 func TestEntitySearch(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 
 	entityIds := []string{
 		"1011d50537fa164b6fef261141797bbe4014526e",
@@ -177,10 +186,7 @@ func TestProcessList(t *testing.T) {
 
 func testProcessList(t *testing.T, procsCount int) {
 	app := vochain.TestBaseApplication(t)
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 
 	// Add 10 entities and process for storing random content
 	for i := 0; i < 10; i++ {
@@ -217,9 +223,8 @@ func testProcessList(t *testing.T, procsCount int) {
 
 	procs := make(map[string]bool)
 	last := 0
-	var list [][]byte
 	for len(procs) < procsCount {
-		list, err = sc.ProcessList(eidTest, last, 10, "", 0, "", "", false)
+		list, err := sc.ProcessList(eidTest, last, 10, "", 0, "", "", false)
 		if err != nil {
 			t.Fatal(err)
 		}
@@ -239,7 +244,7 @@ func testProcessList(t *testing.T, procsCount int) {
 		t.Fatalf("expected %d processes, got %d", procsCount, len(procs))
 	}
 
-	_, err = sc.ProcessList(nil, 0, 64, "", 0, "", "", false)
+	_, err := sc.ProcessList(nil, 0, 64, "", 0, "", "", false)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -247,10 +252,7 @@ func testProcessList(t *testing.T, procsCount int) {
 
 func TestProcessSearch(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 
 	// Add 10 entities and process for storing random content
 	for i := 0; i < 10; i++ {
@@ -398,10 +400,7 @@ func TestProcessSearch(t *testing.T) {
 func TestProcessListWithNamespaceAndStatus(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 
 	// Add 10 processes with different namespaces (from 10 to 20) and status ENDED
 	for i := 0; i < 10; i++ {
@@ -467,13 +466,10 @@ func TestProcessListWithNamespaceAndStatus(t *testing.T) {
 func TestResults(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
 	app.State.SetHeight(3)
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 
 	pid := util.RandomBytes(32)
-	err = app.State.AddProcess(&models.Process{
+	err := app.State.AddProcess(&models.Process{
 		ProcessId:             pid,
 		EnvelopeType:          &models.EnvelopeType{EncryptedVotes: true},
 		Status:                models.ProcessStatus_READY,
@@ -600,10 +596,7 @@ func TestResults(t *testing.T) {
 func TestLiveResults(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 
 	pid := util.RandomBytes(32)
 	if err := app.State.AddProcess(&models.Process{
@@ -671,8 +664,7 @@ func TestLiveResults(t *testing.T) {
 func TestAddVote(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	qt.Assert(t, err, qt.IsNil)
+	sc := newTestScrutinizer(t, app, true)
 
 	options := &models.ProcessVoteOptions{
 		MaxCount:     3,
@@ -780,8 +772,7 @@ func TestBallotProtocolRateProduct(t *testing.T) {
 	// Rate a product from 0 to 4
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	qt.Assert(t, err, qt.IsNil)
+	sc := newTestScrutinizer(t, app, true)
 
 	// Rate 2 products from 0 to 4
 	pid := util.RandomBytes(32)
@@ -817,8 +808,7 @@ func TestBallotProtocolQuadratic(t *testing.T) {
 	// Rate a product from 0 to 4
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	qt.Assert(t, err, qt.IsNil)
+	sc := newTestScrutinizer(t, app, true)
 
 	// Rate 2 products from 0 to 4
 	pid := util.RandomBytes(32)
@@ -868,8 +858,7 @@ func TestBallotProtocolMultiChoice(t *testing.T) {
 
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	qt.Assert(t, err, qt.IsNil)
+	sc := newTestScrutinizer(t, app, true)
 
 	// Rate 2 products from 0 to 4
 	pid := util.RandomBytes(32)
@@ -912,13 +901,10 @@ func TestBallotProtocolMultiChoice(t *testing.T) {
 
 func TestCountVotes(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	if err != nil {
-		t.Fatal(err)
-	}
+	sc := newTestScrutinizer(t, app, true)
 	pid := util.RandomBytes(32)
 
-	err = app.State.AddProcess(&models.Process{
+	err := app.State.AddProcess(&models.Process{
 		ProcessId:    pid,
 		EnvelopeType: &models.EnvelopeType{EncryptedVotes: false},
 		Status:       models.ProcessStatus_READY,
@@ -976,8 +962,7 @@ func TestCountVotes(t *testing.T) {
 func TestTxIndexer(t *testing.T) {
 	app := vochain.TestBaseApplication(t)
 
-	sc, err := NewScrutinizer(t.TempDir(), app, true)
-	qt.Assert(t, err, qt.IsNil)
+	sc := newTestScrutinizer(t, app, true)
 
 	for i := 0; i < 10; i++ {
 		for j := 0; j < 10; j++ {
@@ -985,7 +970,7 @@ func TestTxIndexer(t *testing.T) {
 		}
 	}
 	qt.Assert(t, sc.Commit(0), qt.IsNil)
-	time.Sleep(3 * time.Second)
+	sc.WaitIdle()
 
 	for i := 0; i < 10; i++ {
 		for j := 0; j < 10; j++ {

--- a/vochain/scrutinizer/transaction.go
+++ b/vochain/scrutinizer/transaction.go
@@ -2,6 +2,7 @@ package scrutinizer
 
 import (
 	"fmt"
+	"sync/atomic"
 
 	"github.com/timshannon/badgerhold/v3"
 	"go.vocdoni.io/dvote/log"
@@ -50,8 +51,12 @@ func (s *Scrutinizer) OnNewTx(hash []byte, blockHeight uint32, txIndex int32) {
 // indexNewTxs indexes the txs pending in the newTxPool and updates the transaction count
 // this function should only be called within Commit(), on a new block.
 func (s *Scrutinizer) indexNewTxs(txList []*indexertypes.TxReference) {
+	defer atomic.AddInt64(&s.liveGoroutines, -1)
 	if len(txList) == 0 {
 		return
+	}
+	if s.cancelCtx.Err() != nil {
+		return // closing
 	}
 	s.txIndexLock.Lock()
 	defer s.txIndexLock.Unlock()
@@ -61,6 +66,9 @@ func (s *Scrutinizer) indexNewTxs(txList []*indexertypes.TxReference) {
 		return
 	}
 	for i, tx := range txList {
+		if s.cancelCtx.Err() != nil {
+			return // closing
+		}
 		// Add confirmed txs to transaction count
 		tx.Index = txCount.Count + uint64(i) + 1 // Start indexing at 1
 		if err := s.db.Insert(tx.Index, tx); err != nil {

